core/chains/solana/soltxm/txm.go

```diff
@@ -25,82 +36,341 @@ var (
+		return solanaGo.Signature{}, errors.Wrapf(err, "failed to save tx signature (%s) to inflight txs", sig)
+	}
+
+	// retry with exponential backoff
+	// until context cancelled by timeout or called externally
+	go func() {
+		deltaT := 1 // ms
+		tick := time.After(0)
+		for {
+			select {
+			case <-ctx.Done():
+				// stop sending tx after retry tx ctx times out (does not stop confirmation polling for tx)
+				return
+			case <-tick:
+				retrySig, err := client.SendTx(ctx, tx)
+				// this could occur if endpoint goes down or if ctx cancelled
+				if err != nil {
+					if strings.Contains(err.Error(), "context canceled") {
+						txm.lggr.Debugw("ctx cancelled on send retry transaction", "error", err, "signature", retrySig)
+					} else {
+						txm.lggr.Warnw("failed to send retry transaction", "error", err, "signature", retrySig)
+					}
+
+					break // exit switch
+				}
+				// this should never happen
+				if retrySig != sig {
+					txm.lggr.Criticalw("original signature does not match retry signature", "expectedSignature", sig, "receivedSignature", retrySig)
+				}
+			}
+
+			// exponential increase in wait time, capped at 500ms
+			deltaT *= 2
+			if deltaT > MaxRetryTimeMs {
+				deltaT = MaxRetryTimeMs
+			}
+			tick = time.After(time.Duration(deltaT) * time.Millisecond)
+		}
+	}()
+
+	// return signature for use in simulation
+	return sig, nil
+}
+
+// goroutine that polls to confirm implementation
+// cancels the exponential retry once confirmed
+func (txm *Txm) confirm(ctx context.Context) {
+	defer txm.done.Done()
+
+	timestamps := map[solanaGo.Signature]time.Time{}
+	timestampsMu := sync.Mutex{}
+
+	// if transaction is older than MaxConfirmTimeout, trigger Cancel (removes from list of signatures to check)
+	checkTimestamp := func(sig solanaGo.Signature, warnStr string) {
+		timestampsMu.Lock()
+		timestamp, exists := timestamps[sig]
+		if !exists { // if new, add timestamp
+			timestamps[sig] = time.Now()
+		} else if time.Since(timestamp) > MaxConfirmTimeout { // if timed out, drop tx sig (remove from txs + timestamps)
+			txm.txs.Cancel(sig)
+			txm.lggr.Warnw(warnStr, "signature", sig, "timeoutSeconds", MaxConfirmTimeout)
+			delete(timestamps, sig)
+		}
+		timestampsMu.Unlock()
+	}
+
+	tick := time.After(0)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-tick:
+			// get list of tx signatures to confirm
+			sigs := txm.txs.FetchAndUpdateInflight()
+
+			// skip loop if not txs to confirm
+			if len(sigs) == 0 {
+				break
+			}
+
+			// get client
+			client, err := txm.client.Get()
+			if err != nil {
+				txm.lggr.Errorw("failed to get client in soltxm.confirm", "error", err)
+				break // exit switch
+			}
+
+			// batch sigs no more than MaxSigsToConfirm each
+			sigsBatch := [][]solanaGo.Signature{}
+			for len(sigs) > MaxSigsToConfirm {
+				sigs, sigsBatch = sigs[MaxSigsToConfirm:], append(sigsBatch, sigs[:MaxSigsToConfirm])
+			}
+			sigsBatch = append(sigsBatch, sigs)
```

Comment:
I found this a little difficult to read. WDYT about a separate helper func with comment, or even a small unit test?

Reply:
separated this out into a generic helper func 🎉